[FLINK-25096][runtime] Fixes empty exception history for JobInitializationException

XComp opened the pull request:
## What is the purpose of the change

The `ExecutionGraphInfo` was not initialized properly if a failure happened during the initialization phase. The failure was stored in the `ArchivedExecutionGraph` but not in the `Collection` of `RootExceptionHistoryEntry` elements.

## Brief change log

Instead of providing an empty list, a `RootExceptionHistoryEntry` is added if the passed `ExecutionGraph` contains a failure.

## Verifying this change

I extended the `DefaultJobMasterServiceProcessTest` to cover the `JobInitializationError` case and introduced `ExecutionGraphInfoTest`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

dmvk commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/DefaultJobMasterServiceProcessTest.java:
```diff
@@ -56,7 +63,8 @@
     @Test
-    public void testInitializationFailureCompletesResultFuture() {
+    public void testInitializationFailureCompletesResultFuture()
+            throws ExecutionException, InterruptedException {
```
nit: exceptions are never thrown

dmvk commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/DefaultJobMasterServiceProcessTest.java:
```diff
@@ -70,6 +78,69 @@ public void testInitializationFailureCompletesResultFuture() {
+    }
+
+    @Test
+    public void testInitializationFailureSetsExceptionHistoryProperly()
```
These new two tests are more or less the same. Should we either merge them or have an abstract test that we reuse?

XComp replied:
Good point, I moved the assertions into a helper method. 👍

dmvk commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/DefaultJobMasterServiceProcessTest.java:
```diff
@@ -70,6 +78,69 @@ public void testInitializationFailureCompletesResultFuture() {
+        long afterFailureTimestamp = System.currentTimeMillis();
+
+        final JobManagerRunnerResult result = serviceProcess.getResultFuture().get();
+        final Iterator<RootExceptionHistoryEntry> exceptionHistoryEntryIterator =
+                result.getExecutionGraphInfo().getExceptionHistory().iterator();
+
+        assertTrue(exceptionHistoryEntryIterator.hasNext());
+        final RootExceptionHistoryEntry entry = exceptionHistoryEntryIterator.next();
+        assertFalse(exceptionHistoryEntryIterator.hasNext());
+
+        final Throwable actualException =
+                entry.getException()
+                        .deserializeError(Thread.currentThread().getContextClassLoader());
```
```suggestion
        final Throwable actualException =
                Iterables.getOnlyElement(result.getExecutionGraphInfo().getExceptionHistory())
                        .getException()
                        .deserializeError(Thread.currentThread().getContextClassLoader());
```

dmvk commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/ExecutionGraphInfoTest.java:
```diff
@@ -0,0 +1,72 @@
+        final ErrorInfo failureInfo =
+                executionGraphInfo.getArchivedExecutionGraph().getFailureInfo();
+
+        final Iterator<RootExceptionHistoryEntry> exceptionHistoryEntryIterator =
+                executionGraphInfo.getExceptionHistory().iterator();
+
+        assertTrue(exceptionHistoryEntryIterator.hasNext());
+        final RootExceptionHistoryEntry actualEntry = exceptionHistoryEntryIterator.next();
+        assertFalse(exceptionHistoryEntryIterator.hasNext());
+
```
```suggestion
        final RootExceptionHistoryEntry actualEntry =
                Iterables.getOnlyElement(executionGraphInfo.getExceptionHistory());
```

dmvk reviewed (changes requested):
This fix LGTM overall, great job! 🎉 I have few minor comments on test cosmetics.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/DefaultJobMasterServiceProcessTest.java:
```diff
@@ -63,11 +68,181 @@ public void testInitializationFailureCompletesResultFuture() {
+        CauseTreeContainsAssert.assertThat(initializationFailure).containsCause(originalCause);
+    }
+
+    private static class CauseTreeContainsAssert
```
I migrated to assertj when adding the new tests. I would move the custom Assert implementations along the lines of `flink-test-utils-junit` in `org.apache.flink.core.testutils` along `FlinkMatchers` and deprecate the `FlinkMatchers` class, if you agree. That would enable others to do the same with other tests in the future

dmvk commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/DefaultJobMasterServiceProcessTest.java:
```diff
@@ -63,11 +64,100 @@ public void testInitializationFailureCompletesResultFuture() {
+        final JobManagerRunnerResult actualJobManagerResult =
+                serviceProcess.getResultFuture().join();
+        assertThat(actualJobManagerResult.isInitializationFailure()).isTrue();
+        final Throwable initializationFailure = actualJobManagerResult.getInitializationFailure();
+
+        assertThat(initializationFailure)
+                .satisfies(
+                        t ->
+                                assertThat(
+                                                ExceptionUtils.findThrowable(
+                                                                t, JobInitializationException.class)
+                                                        .isPresent())
+                                        .isTrue());
+        FlinkAsserts.assertThat(initializationFailure)
+                .containsCause(JobInitializationException.class);
+        FlinkAsserts.assertThat(initializationFailure).containsCause(originalCause);
```
```suggestion
        assertThat(actualJobManagerResult.getInitializationFailure())
                .isInstanceOf(JobInitializationException.class)
                .hasCause(originalCause);
```

dmvk commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/DefaultJobMasterServiceProcessTest.java:
```diff
@@ -91,10 +181,10 @@ public void testCloseAfterInitializationSuccess() throws Exception {
-                serviceProcess.getResultFuture(),
-                futureWillCompleteExceptionally(JobNotFinishedException.class, TIMEOUT));
+        assertThat(testingJobMasterService.isClosed()).isTrue();
+        FlinkAsserts.assertThat(serviceProcess.getResultFuture())
+                .failsWithThrowable(TIMEOUT)
+                .containsCause(JobNotFinishedException.class);
```
```suggestion
        assertThat(serviceProcess.getResultFuture())
                .failsWithin(TIMEOUT)
                .withThrowableOfType(ExecutionException.class)
                .withCauseInstanceOf(JobNotFinishedException.class);
```

dmvk commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/DefaultJobMasterServiceProcessTest.java:
```diff
@@ -188,15 +275,17 @@ public void testSuccessOnTerminalState() throws Exception {
-                        .getArchivedExecutionGraph()
-                        .getState(),
-                is(JobStatus.FINISHED));
+        FlinkAsserts.assertThat(serviceProcess.getResultFuture())
+                .succeedsImmediately()
+                .extracting(JobManagerRunnerResult::isSuccess)
+                .isEqualTo(true);
+
+        FlinkAsserts.assertThat(serviceProcess.getResultFuture())
+                .succeedsImmediately()
+                .extracting(JobManagerRunnerResult::getExecutionGraphInfo)
+                .extracting(ExecutionGraphInfo::getArchivedExecutionGraph)
+                .extracting(ArchivedExecutionGraph::getState)
+                .isEqualTo(JobStatus.FINISHED);
```
```suggestion
        assertThat(serviceProcess.getResultFuture())
                .isCompletedWithValueMatching(
                        v ->
                                v.getExecutionGraphInfo().getArchivedExecutionGraph().getState()
                                        == JobStatus.FINISHED);
```

dmvk commented on flink-core/src/test/java/org/apache/flink/testutils/assertj/FlinkAsserts.java:
```diff
@@ -0,0 +1,36 @@
+ */
+public class FlinkAsserts {
+
+    public static CauseTreeContainsAssert assertThat(Throwable actualThrowable) {
+        return new CauseTreeContainsAssert(actualThrowable);
+    }
+
+    public static <T> FlinkCompletableFutureAssert<T> assertThat(CompletableFuture<T> actual) {
+        return new FlinkCompletableFutureAssert<>(actual);
+    }
```
```suggestion
    private static final InstanceOfAssertFactory<Throwable, CauseTreeContainsAssert> CAUSE_TREE_FACTORY =
            new InstanceOfAssertFactory<>(
                    Throwable.class,
                    CauseTreeContainsAssert::new);

    public static InstanceOfAssertFactory<Throwable, CauseTreeContainsAssert> exceptionHierarchy() {
        return CAUSE_TREE_FACTORY;
    }
```

XComp replied:
Good point. I like that. I used a public static field in the implementing class, though, to reduce the boilerplate. WDYT?

dmvk replied:
If we decide to add more "matchers" in the future, would that make them harder to discover? 

I'm thinking along this line: If we'd have all the factories accessible in the same class, would it make developer experience / discovery of the matchers you could leverage smoother?

dmvk commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/DefaultJobMasterServiceProcessTest.java:
```diff
@@ -110,17 +200,14 @@ public void testJobMasterTerminationIsHandled() {
-            assertThat(t, containsCause(RuntimeException.class));
-            assertThat(t, containsMessage(testException.getMessage()));
-        }
+        FlinkAsserts.assertThat(serviceProcess.getResultFuture())
+                .failsImmediately()
+                .containsCause(RuntimeException.class)
+                .containsCauseWithMessage(testException.getMessage());
```
```suggestion
        assertThat(serviceProcess.getResultFuture())
                .failsWithin(TIMEOUT)
                .withThrowableOfType(ExecutionException.class)
                .asInstanceOf(FlinkAsserts.exceptionHierarchy())
                .containsCause(testException);
```

dmvk reviewed (changes requested):
This is a nice progress,I'm really starting to like the new assertj ecosystem.

In general I'm in favor of more sophisticated assertions, however I think we should only provide custom `InstanceOfAssertFactory` factories, that could be nicely chained into the existing assertion ecosystem.

Also I'm not convinced about need for custom asserts for completable futures, the ones provided by assertj are IMO sufficient.

I've added few examples on how the current assert could  be simplified / implemented using existing asserts.

WDYT?

dmvk commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/DefaultJobMasterServiceProcessTest.java:
```diff
@@ -63,11 +64,92 @@ public void testInitializationFailureCompletesResultFuture() {
+
+        assertThat(deserializedException)
+                .isOfAnyClassIn(JobInitializationException.class)
+                .satisfies(
```
could we also leverage the `CauseTreeContainsAssert` here?

XComp replied:
damn, I missed that one. But we can be even more restrictive here and just use the functionality provided by `ThrowableAssert` 👍

dmvk commented on flink-core/src/test/java/org/apache/flink/testutils/assertj/CauseTreeContainsAssert.java:
```diff
@@ -0,0 +1,74 @@
+ * {@code CauseTreeContainsAssert} extends {@link AbstractThrowableAssert} by Flink-specific
+ * assertions.
+ */
+public class CauseTreeContainsAssert
```
Should we come up with a different name? Something more "generic" to the exceptions. If we decide to add new methods in the future, that do different thing than "contains", it could be misleading

XComp replied:
👍 I switched to `FlinkThrowableAssert`

dmvk commented on flink-core/src/test/java/org/apache/flink/testutils/assertj/CauseTreeContainsAssert.java:
```diff
@@ -0,0 +1,74 @@
+        super(actualThrowable, CauseTreeContainsAssert.class);
+    }
+
+    public CauseTreeContainsAssert containsCause(Throwable expectedCause) {
```
Would it be useful to add javadoc to the public methods here?

dmvk commented on flink-core/src/test/java/org/apache/flink/testutils/assertj/CauseTreeContainsAssert.java:
```diff
@@ -0,0 +1,74 @@
+
+    public CauseTreeContainsAssert containsCause(Throwable expectedCause) {
+        return containsCause(
+                "The expected cause " + expectedCause + " was not found in the cause tree.",
```
nit: Would `String.format` make this bit more readable?

dmvk reviewed (commented):
LGTM overall 👍 I've few more questions / comments on the new asserts for exception testing.

Would it make sense (especially for assertj newcomers) to add an example to the documentation on how to use / discover / extend these asserts?

XComp commented:
Although, I'm not that sure anymore whether `FlinkThrowableAssert` wouldn't be the only implementation. `FlinkMatchers` covers only `CompletableFuture` and the cause chain use case.

XComp commented:
@slinkydeveloper I refactored your code after I realized that there was some concurrent development on the same issue. I moved the code into `flink-test-utils-junit` to align it with your approach. I think having this class in `flink-test-utils-junit` makes sense instead of `flink-core` because a few modules benefit from it (even though it means duplicating the `ExceptionUtils` code for traversing the cause chain). Additionally, I realized that it was already the case beforehand (see [FlinkMatchers](https://github.com/apache/flink/blob/master/flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/FlinkMatchers.java#L347-L348)

slinkydeveloper commented on flink-core/src/test/java/org/apache/flink/testutils/assertj/CauseTreeContainsAssert.java:
```diff
@@ -0,0 +1,74 @@
+/*
```
You don't need this class, look at the usages of this: https://github.com/apache/flink/commit/0452506b78db6c19a2e234581f2709d69ce516c6#diff-bfaef4e34a7e5d972ba50489d2b233f781bfff8699b0409fc757ccc11d6fb253R56

slinkydeveloper replied:
You can either use the sample consumers provided or as shown in the javadoc inspect the whole cause chain with the assertions you want.

XComp replied:
I was aware of your change. I was just too stupid to forget pushing my (similar) changes to the remote branch and this PR to add more context to my comment above. I'm fine with using `FlinkAssertions` as it does what it should. The `FlinkThrowableAssert` does the same. 

In the end, it's just unfortunate, that we put effort into the same issue in parallel. ¯\_(ツ)_/¯

XComp replied:
But I did another round of checks and realized that your approach prints a more expressive error message. So, I guess, that's then the reason to switch to your appraoch 👍

slinkydeveloper replied:
Also note that once this gets merged https://github.com/assertj/assertj-core/pull/2423 extracting the chain of causes will look even better!

XComp replied:
That's actually a good idea opening a PR in assertj. I guess this feature might be useful also for other projects. Good job! 👍

dmvk reviewed (approved):
LGTM, good job 🎉 👍 Can you please squash the commit history before merging?
